### PR TITLE
Ldap groups

### PR DESCRIPTION
**Pull request by ghost**
This is rather a big rewrite on the ldap connection in order to add the ability to limit by ldap group.

I'd appreciate some eyes on this looking for anything obviously bad.  There's definitely scope for further refactoring of this, but it appears to work against my test ldap server.

### CODE REVIEW

File: groups.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package main
```
Comment (jacoelho) [inline groups.go:1]: accidentally quadratic 😄

File: groups.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+// return true as soon as we find an item in set that is also in found
+func UserInGroup(set []string, found[]string) bool {
```
Comment (jacoelho) [inline groups.go:4]: this is more like an intersection of slices, not sure if the naming reflects that.

not linted

File: groups.go (hunk 1 above)
Comment (jacoelho) [reply groups.go:1]: do we need a new file?

File: ldapproxy.go [hunk 3]
```diff
@@ -131,15 +133,19 @@ func NewLdapProxy(opts *Options, validator func(string) bool) *LdapProxy {
...
-	ldapConnection := NewLdapConnection(
-		opts.LdapServerHost,
-		opts.LdapServerPort,
-		opts.LdapTLS,
-		opts.LdapBaseDn,
-		opts.LdapBindDn,
-		opts.LdapBindDnPassword,
-		opts.LdapScopeName,
-	)
+	ldapConnection := &ldap.LDAPClient{
```
Comment (jacoelho) [inline ldapproxy.go:136]: needs linting.

you can lint the project: 

```
gofmt -s -w .
```

File: ldapproxy.go [hunk 4]
```diff
@@ -426,11 +443,23 @@ func (p *LdapProxy) SignIn(rw http.ResponseWriter, req *http.Request) {
...
-			http.Redirect(rw, req, redirect, 302)
+			if len(p.LdapGroups) > 0 {
+				log.Printf("User: %s is in groups: %+v", user, groups)
+				if UserInGroup(p.LdapGroups, groups) {
+					session := &SessionState{User: user}
+					p.SaveSession(rw, req, session)
+					http.Redirect(rw, req, redirect, 302)
+				} else {
+					log.Printf("User: %s is not in groups: %+v", user, p.LdapGroups)
+					p.SignInPage(rw, req, 200, true)
```
Comment (jacoelho) [inline ldapproxy.go:456]: should this be a 401 instead?

File: ldapproxy.go [hunk 5]
```diff
@@ -426,11 +443,23 @@ func (p *LdapProxy) SignIn(rw http.ResponseWriter, req *http.Request) {
...
+				}
+			} else {
+				session := &SessionState{User: user}
```
Comment (jacoelho) [inline ldapproxy.go:459]: duplicated with 451. maybe move after the `if ok`

File: ldapproxy.go [hunk 6]
```diff
@@ -426,11 +443,23 @@ func (p *LdapProxy) SignIn(rw http.ResponseWriter, req *http.Request) {
...
-			session := &SessionState{User: user}
-			p.SaveSession(rw, req, session)
-			http.Redirect(rw, req, redirect, 302)
+			if len(p.LdapGroups) > 0 {
+				log.Printf("User: %s is in groups: %+v", user, groups)
+				if UserInGroup(p.LdapGroups, groups) {
+					session := &SessionState{User: user}
+					p.SaveSession(rw, req, session)
+					http.Redirect(rw, req, redirect, 302)
+				} else {
```
Comment (jacoelho) [inline ldapproxy.go:454]: I would remove the else and have a return to keep the flow simple

**Comment by mitchellrj** [comment]
Closed in favour of #4

**Pull request closed without merging**